### ISSUE TITLE
Following the ESLint 9 support, `import/no-unused-modules` only works in flat config if an `.eslintrc` file exists as well

### ISSUE THREAD

**Issue by ronbraha** [import-js/eslint-plugin-import#3079]
I have seen various discussions about this rule, but not quite in the exact same context, so I am unsure if that could be merged with an existing issue: 

I upgraded to `eslint-plugin-import` v2.31.0, which officially supports ESLint V9. Enabling the `import/no-unused-modules` with the existing `unusedExports` option results in the ESLint CLI not recognizing the `eslint.config.js` unless there's a `.eslintrc` file next to it (the `.eslintrc` file can be empty; it just ignores it).

```
Oops! Something went wrong! :(

ESLint: 9.11.1

ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    npm init @eslint/config@latest

ESLint looked for configuration files in <project rootDir> and its ancestors. If it found none, it then looked in your home directory.

If you think you already have a configuration file or if you need more help, please stop by the ESLint Discord server: https://eslint.org/chat
```

As long as the `.eslintrc` exists, the rule works as expected.

**Comment by michaelfaith** [comment]
This is not something that's come up before. There are issues with `no-used-modules` in flat config, but worst case scenario is that it should silently fail (or not ignore things that are specific to ignore).  It shouldn't force you to use an rc config.  We have this rule set up in the flat config example of this repo, if you want to compare your set up with that one.  If that doesn't shed any light on your issue, would you mind sharing your config and the command you're using to run eslint?  We'll need to try and recreate the issue.

**Comment by KristjanTammekivi** [comment]
Hi, can confirm that this is the case. I had a test-repo ready to go, here it is:

https://github.com/KristjanTammekivi/eslint-plugin-import-x-flat-config-debug

```bash
npm run lint # error ESLint couldn't find a configuration file.
touch .eslintrc
npm run lint # linting works, shows unused exports
```

**Comment by ljharb** [comment]
Sounds like we need a dynamic test case to cover this scenario.

**Comment by michaelfaith** [comment]
Thanks for the repo, that's helpful.  It looks like it's blowing up when we're calling the og `FileEnumerator`'s `iterateFiles` function.

https://github.com/import-js/eslint-plugin-import/blob/67cc79841fc823ad4af2532af2dc6704e4b3b03a/src/rules/no-unused-modules.js#L65-L68

So, my first thought is that maybe eslint put more strict controls in place to prevent using that deprecated api in flat config.  However, what's really vexing, is I'm struggling to re-create it in this repo.  We have a flat config example, and even after updating to the latest eslint (the same version as your repo uses), both at the root of the project and in the example project, those same lines of code work fine.  And I double checked that the instance of `FileEnumerator` in both cases is coming from 9.11 eslint in `node_modules` (so it's not some weird multiple installed versions issue).  

![Image](https://github.com/user-attachments/assets/67d45fbd-d555-44b3-8460-2b423b68be79)

I'll keep trying to reproduce it in this repo to either confirm or debunk that initial suspicion.

**Comment by michaelfaith** [comment]
Well damn.  It just occurred to me, while this repo doesn't have an rc config in the example folder I'm running the tests in, it _does_ have one at the root of the repo that it uses for its own linting.  And as we know the way rc config climbs the directory tree to merge configs means it was satisfied by the presence of the root rc config being there.  So this issue was flying under the radar.   At least we know it's not some funky issue with this repo's version set up. 

@ljharb how would you like to approach this?  It doesn't seem as though we can fallback on the deprecated `FileEnumerator` API in v9 after all, and it sounds like there's some issue with moving forward with that original API proposal that we were counting on moving forward.  It seems to be satisfied if the rc file simply exists, but that's a pretty clunky experience.  We could just have it no-op if the flat config is detected, so it doesn't even try to use the old `FileEnumerator`?

**Comment by zoran995** [comment]
@michaelfaith If I got the code correctly it shouldn't even try to use FileEnumerator due to this check https://github.com/import-js/eslint-plugin-import/blob/main/src/rules/no-unused-modules.js#L165-L171, but the session is not available on the context object and it seems it was just available as part of POC created back in February but never got merged into eslint codebase (https://github.com/eslint/eslint/issues/18087#issuecomment-2321443857)

**Comment by michaelfaith** [comment]
Right. It was a bit of a chicken and egg situation. They didn't want to move forward with adding those new functions to the session until we demonstrated demand.  So we added that solution in so that the proposal could move forward (knowing that we'd possibly have to tweak it, if the final api landed in a slightly different form), but still fall back on the deprecated api until it was available. But it seems they've pulled back on that plan (i think).

**Comment by ljharb** [comment]
It'd be better for the rule to error in flat config than silently noop, but even better still to figure out how to get it working.

**Comment by ronbraha** [comment]
@KristjanTammekivi, thanks for the example repo! I saw many discussions around the `no-unused-modules` rule, so I am glad I raised something you were unaware of.
Just for more context: I also experienced this issue in ESLint 8 with a flat config, but I assumed it was a known issue, so I waited for an ESLint 9 support to re-test it

**Comment by michaelfaith** [comment]
> It'd be better for the rule to error in flat config than silently noop, but even better still to figure out how to get it working.

I think there are a couple options that we could explore. We possibly patch `FileEnumerator`. It takes a config file factory as a constructor param 

https://github.com/eslint/eslint/blob/d0a5414c30421e5dbe313790502dbf13b9330fef/lib/cli-engine/file-enumerator.js#L218-L222

That factory class has a `useEslintrc` parameter that defaults to true. 

https://github.com/eslint/eslintrc/blob/fc9837d3c4eb6c8c97bd5594fb72ea95b6e32ab8/lib/cascading-config-array-factory.js#L211

This is where it throw the config error (but only  if `useEslintrc` is true)

https://github.com/eslint/eslintrc/blob/fc9837d3c4eb6c8c97bd5594fb72ea95b6e32ab8/lib/cascading-config-array-factory.js#L519-L522

So that could provide a path to get around it, but honestly i question the value of going that route, since it continues to rely on `FileEnumerator`, which we know will go away in the next major version. So, spending that kind of energy on a temporary band-aid doesn't seem great, _plus_ the `FileEnumerator` solution, when it doesn't hit this issue and the rule runs, it still has the issues with ignored files we found.

Until there's a first-class api added to eslint, like the original planned involved, I don't really see a great option other than throwing a (better) error.  What do you think?

**Comment by ljharb** [comment]
I agree a temporary band-aid isn't great, but it would buy us lots of time because eslint 10 would need to come out *and* it always takes the ecosystem many months to catch up to eslint - so it's not a terrible idea.

If we have a need for a first-class API in eslint, that we can't polyfill/implement ourselves, then that's evidence we'd want to provide to eslint maintainers.

**Comment by michaelfaith** [comment]
> I agree a temporary band-aid isn't great, but it would buy us lots of time because eslint 10 would need to come out _and_ it always takes the ecosystem many months to catch up to eslint - so it's not a terrible idea.

Sounds good.  I can explore that, if nobody else beats me to it.

> If we have a need for a first-class API in eslint, that we can't polyfill/implement ourselves, then that's evidence we'd want to provide to eslint maintainers.
I think your original issue / feature request (https://github.com/eslint/eslint/issues/18087) covered everything.  @nzakas seems unsupportive of covering this use case after the original proposal didn't move forward, and I'm not sure that we have any new information to share?

**Comment by phun-ky** [comment]
So, just to be clear, this plugin is unusable until we have a fix for this? Or does it currently only  require a dummy `.eslintrc` file in the project root, but it will read the `eslint.config.mjs` file?

**Comment by ljharb** [comment]
Not the _plugin_, but this one rule, yes, it seems that's the case.

**Comment by osbornk** [comment]
I just discovered the same case with `import/no-default-esport`

**Comment by guillaumebrunerie** [comment]
Maybe the need of an empty `.eslintrc` should be documented in the mean time?

**Comment by TheJaredWilcurt** [comment]
> `import/no-default-esport`

I set my default esport to [Melee](https://www.youtube.com/watch?v=Ghg0Ex8-DHY).

**Comment by guillaumebrunerie** [comment]
> Maybe the need of an empty `.eslintrc` should be documented in the mean time?

Actually, the `.eslintrc` being still used by the `no-unused-modules` rule to determine which files should be included, it should probably not be empty but instead contain a reasonable `ignorePatterns`. I just checked in my current project, and not having `ignorePatterns` there makes my total lint time go from 15 seconds to 45 seconds.

**Comment by KristjanTammekivi** [comment]
> > I agree a temporary band-aid isn't great, but it would buy us lots of time because eslint 10 would need to come out _and_ it always takes the ecosystem many months to catch up to eslint - so it's not a terrible idea.
> 
> Sounds good. I can explore that, if nobody else beats me to it.
> 
> > If we have a need for a first-class API in eslint, that we can't polyfill/implement ourselves, then that's evidence we'd want to provide to eslint maintainers.
> 
> I think your original issue / feature request ([eslint/eslint#18087](https://github.com/eslint/eslint/issues/18087)) covered everything. [@nzakas](https://github.com/nzakas) seems unsupportive of covering this use case after the original proposal didn't move forward, and I'm not sure that we have any new information to share?

Have You had any time to explore it? (aka bump)

I would do it myself, but I'm not very familiar with either this repo or eslint internals.

**Comment by michaelfaith** [comment]
I finally had some time to work on this this morning: https://github.com/import-js/eslint-plugin-import/pull/3116

Note: This _doesn't_ fix the fact that the `FileEnumerator` doesn't have knowledge of what the user's config is ignoring, it just prevents the rule from looking for a legacy / rc config and erroring out.

**Comment by guillaumebrunerie** [comment]
Wouldn’t it be better to error out with a better error message when using `no-unused-modules` and no `.eslintrc`? The fact that ignore patterns need to be in a `.eslintrc` isn’t obvious at all (especially if it doesn’t error out anymore) and can have a pretty dramatic impact on performance.

**Comment by guillaumebrunerie** [comment]
> it just prevents the rule from looking for a legacy / rc config

Does that mean that you are actually removing support for ignore patterns in the `no-unused-modules` rule when using flat config?

If so, I don’t think that's a good idea at all. The current situation is not ideal (even if you use flat config, you still need to keep a `.eslintrc` with ignore patterns) but at least it works. Removing support for ignore patterns entirely will make this plugin unusable again with flat config in projects that need to ignore build files. For instance, as I mentioned above, this change would make my lint time go from 15 seconds to 45 seconds…

**Comment by michaelfaith** [comment]
> > it just prevents the rule from looking for a legacy / rc config
> 
> Does that mean that you are actually removing support for ignore patterns in the `no-unused-modules` rule when using flat config?

This is true.  I guess I could change it to try again in a catch block if it errors out the first time, so that if you do have an eslintrc it doesn't change the behavior.  There's not really a way to get the old `FileEnumerator` api to be aware of ignores in flat config, without completely re-implementing the ignore logic in `eslint`.  So it's either an eslintrc is required for this rule to know about ignore patterns, or it doesn't try to look for an eslintrc at all (which is what my PR moves it to).  Neither is great, but the later seemed more "correct".  @ljharb thoughts?

**Comment by michaelfaith** [comment]
> I guess I could change it to try again in a catch block if it errors out the first time, so that if you do have an eslintrc it doesn't change the behavior. 

@guillaumebrunerie or are you suggesting we don't even go this far, and instead just accept that eslintrc is required for this rule and log a more expressive error?  If so, i'm ok with that, but then why'd we even wait for me to explore https://github.com/import-js/eslint-plugin-import/issues/3079#issuecomment-2394528538  I made it clear that whatever came of that wouldn't support ignored files.

**Comment by guillaumebrunerie** [comment]
> [@guillaumebrunerie](https://github.com/guillaumebrunerie) or are you suggesting we don't even go this far, and instead just accept that eslintrc is required for this rule and log a more expressive error?

Yes, that's what I would suggest.
It seems to me that if someone wants to use `no-unused-modules`, they probably want to know how to make it ignore files, so it's better to fail if no `.eslintrc` is present with a clear error message like:
```
Due to some restrictions in ESLint 9, the no-unused-modules rule requires a
.eslintrc file to know which files to ignore (even when using flat config).
The .eslintrc file only needs to contains "ignorePatterns", or can be empty
if you do not want to ignore any files.
Check https://github.com/import-js/eslint-plugin-import/issues/3079 for more
context.
```

**Comment by ljharb** [comment]
Hmm. I think that means that #3116 is a breaking change.

Following the above suggestion would indeed prevent a crash and still have it work when an eslintrc is present, but that’s weird and we shouldn’t be aiming for that as a goal state. However, it would be a better intermediate state than the status quo.

**Comment by michaelfaith** [comment]
> Following the above suggestion would indeed prevent a crash and still have it work when an eslintrc is present

It wouldn't really prevent a crash though, if the rc file isn't present, it would just emit a better / more informative error message, right?

**Comment by michaelfaith** [comment]
I updated https://github.com/import-js/eslint-plugin-import/pull/3116 with what we talked about. It no longer attempts to avoid the error and instead throws an error with a more meaningful message if flat config is enabled and `FileEnumerator` errored out due to a lack of rc file.

![Image](https://github.com/user-attachments/assets/42b517b2-22bc-4033-ae2e-df072621ce41)

**Comment by guillaumebrunerie** [comment]
> Can we get a release, this issue has been preventing me from upgrading to ESLint v9 across my repos (personal and work) for months. The issue is resolved, so we just need a build/release! :)

For me the latest version of `eslint-plugin-import` has been working fine for months with ESLint v9. You just need to keep a `.eslintrc` containing `ignorePatterns` in order for `import/no-unused-modules` to work.

**Comment by TheJaredWilcurt** [comment]
I'm trying to avoid rework across a bunch of repos to remove a random file.  It also complicates the instructions I'm writing for other teams at work as to how to adopt ESLint 9. I'd like to create a dynamic site to generate a config file ([prototype example](https://jsfiddle.net/yder40ap/)). And I'd like to avoid adding in a rule that disables `import/no-unused-modules` with a comment explaining you either need `eslint-plugin-import` at `>2.31.0` or to create an empty `.eslintrc` file because of a bug that's been fixed.

**Comment by guillaumebrunerie** [comment]
> I'd like to avoid adding in a rule that disables `import/no-unused-modules` with a comment explaining you either need `eslint-plugin-import` at `>2.31.0` or to create an empty `.eslintrc` file because of a bug that's been fixed.

Well, then it hasn't been fixed, it still requires a `.eslintrc` file because the only API that `import/no-unused-modules` can use is an old API that has no knowledge of flat config.

**Comment by ljharb** [comment]
@TheJaredWilcurt i'll cut a release as soon as I have time, and also as soon as it's ready, which it still isn't. Please be patient.

**Comment by phun-ky** [comment]
@ljharb suggest that this issue is reopened until the fix is released, right?

**Comment by ljharb** [comment]
@phun-ky no, that’s not how GitHub works - issues are closed when fixes are merged, unrelated to whether the fixes are released.

**Comment by phun-ky** [comment]
🤔 If it's fixed, it's usable and in production IMHO.. Which currently this fix is not usable or in production. In my experience that is. That kind of process is common on larger scale projects. But I see more and more of issues closes when it is released and verified.

I will patiently wait for the release. Appreciate the good work! <3

**Comment by guillaumebrunerie** [comment]
@ljharb Although in this case, the issue in the title has *not* been fixed. `import/no-unused-modules` still only works in flat config if a `.eslintrc` file exists as well, it's only that the error message has been improved (it explains the need of a `.eslintrc` file rather than crashing unexpectedly).

@phun-ky Unless it's particularly problematic for you to just add an empty `.eslintrc` file, the latest stable version of `eslint-plugin-import` works perfectly fine with flat config and the latest ESLint.

**Comment by ljharb** [comment]
@guillaumebrunerie that's fair, and worth reopening for.

**Comment by OnkelTem** [comment]
All this time one could kindly update the docs. Really. 

P.S. It took two hours for me to find out that `'import/no-unused-modules'` was the culprit here and only then I could find this issue and create empty **.eslintrc** which obviously resolved the problem. And this is not some edge case, but the very mainstream one on the current ESLint 9. Meh.

Anyway, thanks for you work.  Look forward to seeing it released.

**Comment by TheJaredWilcurt** [comment]
Yeah, I finally caved and added a giant purple banner to my config automator to warn about this to my users so we can move on to ESLint 9 upgrades:

* https://tjw-lint.github.io/?checked=[tjwImport]

**Comment by michaelfaith** [comment]
> All this time one could kindly update the docs. Really.

This repo accepts PRs, if you'd like to put one in.
